### ISSUE TITLE
Ability to stop a service

### ISSUE THREAD

**Issue by mashwino** [kdebisschop/rundeck-rancher-node-plugin#20]
Hi,
thank you for sharing this plugin, very very useful when you have a Rancher infrastructure to manage.
In certain case we need to stop a service before upgrading an other, is there a way to achieve this ?
Thank you.

**Comment by kdebisschop** [comment]
I'm definitely seeking collaboration both in defining a useful and organized feature set, and in code quality and review. So I am very glad to see the plugin get used by others.

To be sure I understand the request, additional node steps to stop and to start a service would be required? If so, those do not yet exist but would not be hard to write.

**Comment by kdebisschop** [comment]
Hmmm - stop is available for containers, but not for services. So I guess this would need to be "stop all containers in the service." Another complication is Rancher may restart the containers, depending on your configuration. But maybe it is OK to assume that is handled outside the plugin framework.

**Comment by kdebisschop** [comment]
Or maybe Activate/Deactivate/Restart?

**Comment by mashwino** [comment]
Hi,
thanks a lot for your time, let me precise my case, in a process of upgrading a stack, the first thing I need to do is stoping a specific service (and all containers attached to it), before using a "only once" container to upgrade database, if I do not do so, risk is high to get locks on database and the process may not end well. Activate/Deactivate/Restart seems to me the best choice.

Have a nice day.
++

**Comment by kdebisschop** [comment]
Please try and keep one issue per issue - I have created another issue for the bug in the file copier. and I will create another for the feature request.

**Comment by kdebisschop** [comment]
Stopping (deactivating) works.... But once stopped the are no nodes for Rancher to access. Unless I can think of a workaround, this feature may require a new resource plugin that is based on services rather than containers.

**Comment by kdebisschop** [comment]
I am thinking of altering the resource so you can choose containers, services, or both as nodes. Then each node would need to indicate whether it is a service or container, and each plugin would need to be able to understand that. Unfortunately, file copy and exec are fundamentally things that happen on a container, so i do not think just switching to services as the definition of a node is the correct fix. So it will be extra code to provide a uniform way to get from container to its service and from a service to one or all of its containers.

**Comment by mashwino** [comment]
Hello,
I understand the problem, I tried to upgrade a stopped service and got the error message, is it a limitation from rancher api or from the upgrade plugin to have the service state "running" ?
May be having the ability at least to deactivate a service could do the trick even if to reactivate it we will need to find an another way. Thank you any way, I removed the precedent remarks from my last post and opened a new issue.
Have a nice day.

**Comment by kdebisschop** [comment]
Actually, those are separate problems. Yes, Rancher does not allow you to upgrade a degraded service, but that is not the problem I face here. The problem here is that the plugin currently defines nodes as containers. When the service is deactivated, there are no containers to identify as nodes so the service cannot be reactivated. This problem will go away if the nodes can be defined as services instead of containers. But file copy and exec are inherently container-based processes, so defining nodes as services does not work across the board either.

**Comment by kdebisschop** [comment]
I want to avoid big untested changes. So I have a plan that does this in steps:

1. Add the ability to select services as nodes. During this phase, NodeExecutor and FileCopier would only work for containers.
2. Add a Node Step Plugin to manage a service (Activate/Deactivate/Restart)

I am going to commit that work soon.

The next step will be to extend NodeExecutor and FileCopier to work for all containers in a service when the selected node is a service instead of a single container. That will be a separate feature.

**Comment by mashwino** [comment]
Hi,
seems exactly what we would need, perfect. :-)
Thank you for your time.
